### DISCUSSION TITLE
Extracting User provided information from PDF - how to do so?

### DISCUSSION THREAD

**Discussion by mangowithice** [pymupdf/PyMuPDF#2862] [Q&A]
hiya - i have a PDF written in Thai, and there are empty space for users to provide information via Fill & Comment in Acrobat PDF. When I extract text from the PDF, the provided text will be consolidated towards the end of the output dump (i.e. PDF original text displayed first, followed by the user provided text). if i do a SORT=TRUE, the order of the provided text will be random. My aim is to extract the provided information and store them . 

I guessed I can find out the coordinates of the text boxes where user potentially fill and write the function to detect the text based on the bbox. But because this is not a proper form per-se, if the input overflow out of the box, i hit an issue.

Is there a way for pymupdf to extract the text in some logic sequence? The reason i am exploring pymupdf is because it can print Thai much better than tabula and pdfplumber, However I am unable to find the most appropriate approach to detect user text.

The highlighted below in yellow are what user keyed in. 
![image](https://github.com/pymupdf/PyMuPDF/assets/152759078/653a82cb-6b8e-44c3-b2b4-a385fdf3aeb2)

**Comment by JorjMcKie** [comment]
I am afraid there is no way other than developing your own logic that fits user input in the right places.
That input indeed is stored after the pages "standard" content part - this is inevitable isn't it?

So you must sort the extracted text in some clever way.
As far as I can see from the picture, the user text has a baseline a little higher up than the preceding (key) word. Another complexity is that you have those underscores "_" which do not belong to the text you are interested in - but probably also are extracted, right?
I suggest you do this:
1. Extract the words of the page via `words = page.get_text("words", sort=True, delimiters="_")`. This will omit all white space **_and_** underscores. The word items are sorted by ascending y, then x coordinate.
2. Define a list into which you will store all words belonging to the same (refactored) line.
3. Walk through the words and put all words in the same line item (your refactored line list) if their bottom (y1) coordinate is in vertical range of the other words in that line list item.
4. for each refactored line item, sort its words by x0 coordinates and join them with one space between any two.

That should ultimately give you the desired output.

**Reply by JorjMcKie** [reply]
If you want, let me have an example page and I can better explain, how this works.
